javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedSource.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @name Inclusion of untrusted functionality by a HTML element.
+ * @description Including untrusted functionality by a HTML element
+ *              opens up for potential man-in-the-middle attacks.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 8.1
+ * @precision high
+ * @id js/functionality-from-untrusted-source
+ * @tags security
+ *       external/cwe/cwe-830
+ */
+
+import javascript
+import semmle.javascript.HTML
+
+bindingset[host]
+predicate isAllowedHost(string host) { host.toLowerCase().regexpMatch("localhost(:[0-9]+)?/.*") }
+
+bindingset[path]
+predicate isUntrustedSourcePath(string path) {
+  path.substring(0, 2) = "//"
+  or
+  exists(string hostPath | hostPath = path.regexpCapture("http://(.*)", 1) |
+    not isAllowedHost(hostPath)
+  )
+}
+
+abstract class IncludesUntrustedContent extends HTML::Element {
+  IncludesUntrustedContent() { this = this }
```

Comment:
I would hope we can just delete this tautological charpred.
```suggestion
```

Reply:
There I was, thinking that abstract classes really need a characteristic predicates. That was a wrong assumption :) thanks for pointing it out.